### PR TITLE
3438 implement missing tests for sequencer

### LINKED ISSUE #3438: implement missing tests for Sequencer
## Motivation

**Current architecture**

some edge cases are not covered by tests

- **Critical Problem (hot)**: 
  - unfinished
  - edge cases not covered 

## Technical design
- [x] https://github.com/voedger/voedger/blob/2629a77d6998469ca19f3bd3928a52dfa7eb9d82/pkg/isequencer/impl_test.go#L63
  - eliminate this, the storage must not decide whether update or not. It must just write averything provided.
- [x] retry_tests:
  - eliminate sleep in test funcs. There must not be any sleep except timeout test
  - implement timeout test
  - check coverage: branches that *could not* be tested must be marked as `// notest` comment
- [x] https://github.com/voedger/voedger/blob/2629a77d6998469ca19f3bd3928a52dfa7eb9d82/pkg/isequencer/impl_test.go#L88
  - implement `mockStorage.ActualizeSequencesFromPLog`
- ~do not use `Params.SeqTypes` as the stroage in tests. All `Number`s in `Params.SeqTypes`must be 1~
  - must be used e.g. `sitructs.FirstBaseRecordID`
- ~must `cleanup()` flush all remaining first, @maxim-ge ?~
  - no
- [ ] implement a complex test like this:
  - provide a storage that already contains 4 events:
    - 1
      - plogoffset 1
      - []SeqValue is empty
    - 2
      - plogoffset 2
      - []SeqValue{{Key: WSID 1, SeqID 1, Value: 1}}
    - 3
      - plogoffset 3
      - []SeqValue{{Key: WSID 1, SeqID 1, Value: 2}, {Key: WSID 1, SeqID 2, Value: 1}, {Key: WSID 1, SeqID 3, Value: 4}, {Key: WSID 1, SeqID 3, Value: 3}} 
    - 4
      - plogoffset 4
      - []SeqValue{{Key: WSID 2, SeqID 1, Value: 42}, {Key: WSID 2, SeqID 2, Value: 41}, {Key: WSID 2, SeqID 3, Value: 44}, {Key: WSID 2, SeqID 3, Value: 43}} 
  - test for Next + flush: 4 cases:
    - test taking the next from storage + cache: use Flush() as `commit`
    - test taking the next from storage + incproc: use (Flush() + purge cache) as `commit`
    - test taking the next from storage + cache using timeout instead of Flush: itime.Sleep(flushTimeout) as `commit`
    - test taking the next from storage + incproc using timeout instead of Flush:: use (itime.Sleep(flushTimeout) + purge cache) as `commit`
      - run the following 2 times: first expect values as written below, 2nd run - +1 to each
        - New()
        - Start(wsKind, wsid 1)  - expect plogoffset 5
        - Next(seqID 1) - expect 3
        - Next(seqID 2) - expect 2
        - Next(seqID 3) - expect 5
        - Next(seqID 3) - expect 1
        - `commit`
        - Start(wsKind, wsid 1)  - expect plogoffset 6
        - Next(seqID 1) - expect 3
        - Next(seqID 2) - expect 2
        - Next(seqID 3) - expect 5
        - Next(seqID 4) - expect 2
        - `commit`
        - Start(wsKind, wsid 2)  - expect plogoffset 7
        -  Next(seqID 1) - expect 43
        -  Next(seqID 2) - expect 42
        -  Next(seqID 3) - expect 45
        -  Next(seqID 4) - expect 1
        - `commit`
        - clenaup()
  - test for Next + Actualize
    - run the following 2 times, on each expect the same numbers
      - New()
      - Start(wsKind, wsid 1)  - expect plogoffset 5
      - Next(seqID 1) - expect 3
      - Next(seqID 2) - expect 2
      - Next(seqID 3) - expect 5
      - Next(seqID 4) - expect 1
      - Actualize()
      - Start(wsKind, wsid 1)  - expect plogoffset 5
      - Next(seqID 1) - expect 3
      - Next(seqID 2) - expect 2
      - Next(seqID 3) - expect 5
      - Next(seqID 4) - expect 1
      - Actualize() 
      - Start(wsKind, wsid 2)  - expect plogoffset 5
      -  Next(seqID 1) - expect 43
      -  Next(seqID 2) - expect 42
      -  Next(seqID 3) - expect 45
      -  Next(seqID 4) - expect 1
      - Actualize() 
      - clenaup()

### PR DESCRIPTION
**Pull request by anurmanov**
fix #3438

### CODE REVIEW

**Review by host6 — CHANGES_REQUESTED** [review]

File: pkg/isequencer/types.go [hunk 1]
```diff
@@ -99,6 +100,154 @@ type sequencer struct {
...
 
 	iTime coreutils.ITime
 }
+
+// MockStorage implements ISeqStorage for testing purposes
+type MockStorage struct {
+	mu                        sync.RWMutex
+	Numbers                   map[WSID]map[SeqID]Number
+	NextOffset                PLogOffset
+	pLog                      map[PLogOffset][]SeqValue // Simulated PLog entries
```
[the remark this answers is not in the document]
Comment (anurmanov) [reply pkg/isequencer/types.go:114]: what about ordered by offset list?

[the remark this answers is not in the document]
Comment (host6) [reply pkg/isequencer/types.go:114]: wrong way. Will modify ActualizeFromPLog() func instead

File: pkg/isequencer/types.go [hunk 2]
```diff
@@ -99,6 +100,163 @@ type sequencer struct {
...
+}
+
+// ActualizeSequencesFromPLog implements isequencer.ISeqStorage.ActualizeSequencesFromPLog
+func (m *MockStorage) ActualizeSequencesFromPLog(ctx context.Context, offset PLogOffset, batcher func(ctx context.Context, batch []SeqValue, offset PLogOffset) error) error {
+	// notest
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	// Process entries in the mocked PLog from the provided offset
+	for i, batch := range m.pLog {
```
Comment (host6) [inline pkg/isequencer/types.go:222]: should run through m.pLog this way:
```golang
for offsetProbe := offset; offsetProbe < math.MaxInt;offsetProbe++ {
  batch, ok := m.pLog[offsetProbe ]
  if !ok {
    // end of PLog is reached
    return
  }
  ...
```

File: pkg/isequencer/impl.go [hunk 3]
```diff
@@ -265,31 +218,32 @@ func (s *sequencer) Next(seqID SeqID) (num Number, err error) {
...
 	if err != nil {
+		// happens when ctx is closed during storage error
 		return 0, err
 	}
 
 	// If number is 0 then initial value is used
 	nextNumber = knownNumbers[0]
 	if nextNumber == 0 {
-		nextNumber = initialValue
+		nextNumber = initialValue - 1 // initial value 1 and there are no such records in plog at all -> should issue 1, not 2
```
[the remark this answers is not in the document]
Comment (host6) [reply pkg/isequencer/impl.go:243]: fixed by checking: initial numbers can no be <1 on New()

**Review by host6 — COMMENTED** [review]
go on

**Review by host6 — APPROVED** [review]
